dokka-integration-tests/gradle/src/main/kotlin/org/jetbrains/dokka/it/gradle/gradleCCReportData.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package org.jetbrains.dokka.it.gradle
+
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+import org.jetbrains.dokka.gradle.utils.GradleProjectTest
+import java.nio.file.Path
+import kotlin.io.path.isRegularFile
+import kotlin.io.path.readText
+import kotlin.io.path.walk
+
+/**
+ * Gradle creates an HTML report for each Configuration Cache result.
+ * This report contains JSON data that we can parse to get fine-grained details about the
+ * Configuration Cache result.
+ *
+ * Only the first CC report will be parsed, so make sure to clean the CC report directory
+ * before running any Gradle tasks to ensure only one report is found.
+ */
+fun GradleProjectTest.loadConfigurationCacheReportData(): ConfigurationCacheReportData? {
+    val ccReportFile = projectDir.resolve("build/reports/configuration-cache")
+        .walk()
+        .filter { it.isRegularFile() }
+        .singleOrNull()
+        ?: return null
+
+    return parseCCReportData(ccReportFile)
+}
+
+@Serializable
+data class ConfigurationCacheReportData(
+    val diagnostics: List<DiagnosticsItem>,
+    val totalProblemCount: Int,
+    val buildName: String,
+    val requestedTasks: String,
+    val cacheAction: String,
+    val documentationLink: String,
+) {
+
+    @Serializable
+    data class DiagnosticsItem(
+        val trace: List<Trace>,
+        val input: List<Input>,
+        val documentationLink: String? = null,
+    )
+
+    @Serializable
+    data class Trace(
+        val kind: String,
+        val type: String? = null,
+        val location: String? = null,
+    )
+
+    @Serializable
+    data class Input(
+        val text: String? = null,
+        val name: String? = null,
+    )
+}
+
+/**
+ * Extract and parse the JSON data from [ccReport].
+ */
+private fun parseCCReportData(ccReport: Path): ConfigurationCacheReportData {
+    val reportData = ccReport.readText()
+        .substringAfter("// begin-report-data", "could not find 'begin-report-data'")
+        .substringBefore("// end-report-data", "could not find 'end-report-data'")
+
+    return Json.decodeFromString<ConfigurationCacheReportData>(reportData)
```

Comment:
Looking at this CC check I have a couple of thoughts:
* Could we just `ignoreUnknownKeys` and only read the [totalProblemCount](https://github.com/Kotlin/dokka/pull/3821/files#diff-e60a0378c359ab13552c8a4083fb4f95be9a4b66f36d7d4cbf4419e8707600f1R332)? Otherwise this will break if the CC report json format changes
* Do we really want to parse this json at all? What is the benefit compared to just failing the test?
* Does this CC check ever actually run for a build that fails due to a CC serialization error? I think the test is going to fail prematurely with a build failure or an assertion failure before the `totalProblemCount` assertion runs

Reply:
> * Could we just `ignoreUnknownKeys` and only read the [totalProblemCount](https://github.com/Kotlin/dokka/pull/3821/files#diff-e60a0378c359ab13552c8a4083fb4f95be9a4b66f36d7d4cbf4419e8707600f1R332)? Otherwise this will break if the CC report json format changes

Yes, we could add `ignoreUnknownKeys`, but at the moment it's simple to parse everything. We could add it in the future if necessary.

The CC test here is a 'soft launch'. In the future I want to specifically read the JSON CC report data and test the registered inputs are as expected. 

<img width="600" alt="image" src="https://github.com/user-attachments/assets/94594059-abf7-42c9-a9c4-d66084c0fe9d">

> * Do we really want to parse this json at all? What is the benefit compared to just failing the test?

The difference with checking the CC report JSON is we can be more certain if CC fails. Just parsing the console output can lead to false positives and uncertain results. We can also get and report more specific information about failures, which might not be present in the console output.

> * Does this CC check ever actually run for a build that fails due to a CC serialization error? I think the test is going to fail prematurely with a build failure or an assertion failure before the `totalProblemCount` assertion runs

iirc there are some CC problems that are reported, but don't fail the build. But I can't remember them right now...

Reply:
For me this check with JSON parsing looks not so useful, as the build should definitely fail, if CC is not supported for some reason with CC enabled.
I feel like it brings more harm (maintainability, possible format changes, etc) than profits (which are unknown to me).
E.g ` We can also get and report more specific information about failures, which might not be present in the console output.` It's always possible to run the test locally, anyway, to fix/investigate the issue we will need it. Of course it's not an option for TC, but I think that we could just then store CC reports somewhere - IMO, this will be easier to maintain and will have more profit (as we could read the whole report)

Reply:
>  there are some CC problems that are reported, but don't fail the build

Are these going to be reflected in the `totalProblemCount`? In case they are, I still wonder if specifying `org.gradle.configuration-cache.problems=fail` and `org.gradle.configuration-cache.max-problems=0` will achieve  failing the build with CC errors

> report more specific information about failures

Would it be possible to zip the entire `build/reports/configuration-cache` directory into the GH/TC artifacts to be able to inspect the full Gradle report?

> test the registered inputs are as expected.

I want to highlight that this might bring a maintenance cost where some CC inputs are going to change due to changes like KGP version, which are unlikely to be of interest to Dokka.